### ISSUE TITLE
Is React Native at all supported?

### ISSUE THREAD

**Issue by kuka** [pubkey/rxdb#2975]
## Case
Clarification

## Issue
React Native adapters aren’t in a good shape. Is it all supported as docs say it is? I played with several possible solutions (incl. libs by @craftdogz) and it doesn’t look that good. Has anyone managed to actually get it to work? I have a setup for web and electron that works flawlessly but React Native’s been a disaster so far. Just want to know what I’m dealing with before investing more time into it. Thanks!

## Info
  - Environment: React Native
  - Adapter: AsyncStorage/SQLite
  - Stack: JS/React

**Comment by kuka** [comment]
Well, the good news is that I found a solution using @craftdogz libs to get the RxDB to work with the SQLite PouchDB in React Native with Hasura with full replication. I’ll see how well it works but it’s good enough for now. I’ll close this issue. If anyone has any questions or wants to discuss alternatives for storage in React Native env, you’re welcome to do so.

**Comment by calebgregory** [comment]
@kuka congratulations on getting it working!  any hints how you managed to do so?

**Comment by kuka** [comment]
@calebgregory Thanks mate! I pretty much used @craftzdog’s libs (https://github.com/craftzdog/pouchdb-react-native, https://github.com/craftzdog/pouchdb-adapter-react-native-sqlite) to accomplish it.

I use Yarn Workspaces + Learna. This is the snippet from my `Database.js` in `packages/mobile`:

```js
import SQLite from 'react-native-sqlite-2'
import SQLiteAdapterFactory from 'pouchdb-adapter-react-native-sqlite'

import { createRxDatabase, removeRxDatabase, addRxPlugin } from 'rxdb'
import { RxDBReplicationGraphQLPlugin } from 'rxdb/plugins/replication-graphql'

const SQLiteAdapter = SQLiteAdapterFactory(SQLite)

addRxPlugin(SQLiteAdapter)
addRxPlugin(require('pouchdb-adapter-http'))
addRxPlugin(RxDBReplicationGraphQLPlugin)

const db = await createRxDatabase({
  name,
  adapter: 'react-native-sqlite',
  multiInstance: false,
  eventReduce: false
})
```

`babel.config.js` updates to resolve forks:

```js
module.exports = function exporting(api) {
  api.cache.never()
  return {
    sourceMaps: true,
    presets: [
      'module:metro-react-native-babel-preset'
    ],
    plugins: [
      [
        'module-resolver',
        {
          alias: {
            'pouchdb-md5': 'react-native-pouchdb-md5',
            'pouchdb-binary-utils':
              '@craftzdog/pouchdb-binary-utils-react-native',
          },
        },
      ],
    ]
  }
}

```

These are the main deps from `packages/mobile/package.json`:

```js
    "@craftzdog/pouchdb-binary-utils-react-native": "7.2.4",
    "@craftzdog/pouchdb-core-react-native": "7.2.2",
    "@react-navigation/bottom-tabs": "5.11.8",
    "@react-navigation/drawer": "5.12.4",
    "@react-navigation/native": "5.9.3",
    "base-64": "1.0.0",
    "events": "3.3.0",
    "invariant": "2.2.4",
    "pouchdb-adapter-http": "7.2.2",
    "pouchdb-adapter-react-native-sqlite": "3.0.1",
    "pouchdb-mapreduce": "7.2.2",
    "pouchdb-replication": "7.2.2",
    "process": "0.11.10",
    "react": "17.0.1",
    "shared": "file:../shared",
    "react-native": "0.64.0",
    "react-native-get-random-values": "1.6.0",
    "react-native-pouchdb-md5": "7.3.0",
    "react-native-quick-md5": "1.0.2",
    "react-native-sqlite-2": "3.2.0",
    "rxdb": "9.15.0",
    "rxjs": "6.6.6"
```

**Comment by calebgregory** [comment]
@kuka you da you da best!  thank you!!

**Comment by rush86999** [comment]
i don't think you need rxjs do you?

**Comment by rush86999** [comment]
@kuka why is eventReduce false? Is there a conflict issue with sqlite?

**Comment by kuka** [comment]
> @kuka why is eventReduce false? Is there a conflict issue with sqlite?

Can’t say now, I switched to WatermelonDB.

**Comment by hdwatts** [comment]
@kuka how do you like the switch? I finally got my app in a working state on web and mobile but am considering switching over to WatermelonDB just because of how tenuous it feels.

**Comment by kuka** [comment]
@hdwatts Well, I now have a Dexie on the web and WatermelonDB on the RN. I can't say I love the stack, but the performance is the key this time around, so I chose to stick with these, and so far, so good. Dexie is great, but it doesn't have an RN-compatible solution that is stable enough for production. WatermelonDB is great performance-wise. Sync is possible too.

When I started working on this, I thought there'd be a more-or-less widely used solution, but there are many edge cases when it comes to local/offline-first codebases. PouchDB isn't bad, but some drawbacks weren't good in my project as it has a lot of nested data (as in "data.layout.xs[3].blocks[2].size"), and syncing those in chunks weren't feasible. Now I have a uWebSockets-based push/pull and Y.js (CRDT) docs that fit the use case. We'll see how it works in the long run.

If you hit a similar bottleneck with the RxDB, then giving a try isn’t a bad idea. Depends on how complex your data is. I’m making a text/layout editor so nested data is not avoidable in the long run.
